hbase-server/src/main/java/org/apache/hadoop/hbase/wal/WALFactory.java

```diff
@@ -260,14 +264,27 @@ private WALFactory(Configuration conf, String factoryId, Abortable abortable) th
     // end required early initialization
     if (conf.getBoolean(WAL_ENABLED, true)) {
       WALProvider provider = createProvider(getProviderClass(WAL_PROVIDER, DEFAULT_WAL_PROVIDER));
+      WALProvider providerInReplicationScope =
+        createProvider(getProviderClass(WAL_PROVIDER, DEFAULT_WAL_PROVIDER));
       provider.init(this, conf, null, this.abortable);
       provider.addWALActionsListener(new MetricsWAL());
       this.provider = provider;
+
+      if (conf.getBoolean(REPLICATION_WAL_FILTER_BY_SCOPE_ENABLED,
+          REPLICATION_WAL_FILTER_BY_SCOPE_ENABLED_DEFAULT)) {
+        providerInReplicationScope.init(this, conf,
+          AbstractFSWALProvider.REPLICATION_WAL_PROVIDER_ID, this.abortable);
+        providerInReplicationScope.addWALActionsListener(new MetricsWAL());
+        this.providerInReplicationScope = providerInReplicationScope;
+      } else {
+        this.providerInReplicationScope = null;
+      }
     } else {
       // special handling of existing configuration behavior.
       LOG.warn("Running with WAL disabled.");
       provider = new DisabledWALProvider();
       provider.init(this, conf, factoryId, null);
+      providerInReplicationScope = null;
```

Comment:
nit: do we really need setting null here?

Reply:
Ah, if we set null here then if we turn on this feature but not enable WAL there will be an NPE. Has changed.